# Import Libraries

In [104]:
import numpy as np
import pandas as pd

# Loading the Data and Selecting the Features

### Import Data

In [107]:
loan_data_inputs_train = pd.read_csv('loan_data_inputs_train.csv', index_col = 0)
loan_data_targets_train = pd.read_csv('loan_data_targets_train.csv', index_col = 0)
loan_data_inputs_test = pd.read_csv('loan_data_inputs_test.csv', index_col = 0)
loan_data_targets_test = pd.read_csv('loan_data_targets_test.csv', index_col = 0)

### Explore Data

In [109]:
loan_data_inputs_train.head()

Unnamed: 0        id  member_id  loan_amnt  funded_amnt  \
427211      427211  12796369   14818505      24000        24000   
206088      206088   1439740    1691948      10000        10000   
136020      136020   5214749    6556909      20425        20425   
412305      412305  13827698   15890016      17200        17200   
36159        36159    422455     496525       8400         8400   

        funded_amnt_inv        term  int_rate  installment grade  ...  \
427211          24000.0   36 months      8.90       762.08     A  ...   
206088          10000.0   36 months     14.33       343.39     C  ...   
136020          20425.0   36 months      8.90       648.56     A  ...   
412305          17200.0   36 months     16.59       609.73     D  ...   
36159            7450.0   36 months     12.84       282.40     C  ...   

       dti:21.7-22.4 dti:22.4-35 dti:>35 mths_since_last_record:Missing  \
427211             1           0       0                              1   
206088             0           0       0                              1   
136020             0           0       0                              1   
412305             0           0       0                              1   
36159              0           1       0                              1   

        mths_since_last_record:0-2 mths_since_last_record:3-20  \
427211                           0                           0   
206088                           0                           0   
136020                           0                           0   
412305                           0                           0   
36159                            0                           0   

       mths_since_last_record:21-31 mths_since_last_record:32-80  \
427211                            0                            0   
206088                            0                            0   
136020                            0                            0   
412305                            0                            0   
36159                             0                            0   

       mths_since_last_record:81-86 mths_since_last_record:>86  
427211                            0                          0  
206088                            0                          0  
136020                            0                          0  
412305                            0                          0  
36159                             0                          0  

[5 rows x 324 columns]

In [110]:
loan_data_targets_train.head()

good_bad
427211         1
206088         1
136020         1
412305         0
36159          0

In [111]:
loan_data_inputs_train.shape

(373028, 324)

In [112]:
loan_data_targets_train.shape

(373028, 1)

In [113]:
loan_data_inputs_test.shape

(93257, 324)

In [114]:
loan_data_targets_test.shape

(93257, 1)

### Selecting the Features

In [116]:
# Here we select a limited set of input variables in a new dataframe.
inputs_train_with_ref_cat = loan_data_inputs_train.loc[: , ['grade:A',
'grade:B',
'grade:C',
'grade:D',
'grade:E',
'grade:F',
'grade:G',
'home_ownership:RENT_OTHER_NONE_ANY',
'home_ownership:OWN',
'home_ownership:MORTGAGE',
'addr_state:ND_NE_IA_NV_FL_HI_AL',
'addr_state:NM_VA',
'addr_state:NY',
'addr_state:OK_TN_MO_LA_MD_NC',
'addr_state:CA',
'addr_state:UT_KY_AZ_NJ',
'addr_state:AR_MI_PA_OH_MN',
'addr_state:RI_MA_DE_SD_IN',
'addr_state:GA_WA_OR',
'addr_state:WI_MT',
'addr_state:TX',
'addr_state:IL_CT',
'addr_state:KS_SC_CO_VT_AK_MS',
'addr_state:WV_NH_WY_DC_ME_ID',
'verification_status:Not Verified',
'verification_status:Source Verified',
'verification_status:Verified',
'purpose:educ__sm_b__wedd__ren_en__mov__house',
'purpose:credit_card',
'purpose:debt_consolidation',
'purpose:oth__med__vacation',
'purpose:major_purch__car__home_impr',
'initial_list_status:f',
'initial_list_status:w',
'term:36',
'term:60',
'emp_length:0',
'emp_length:1',
'emp_length:2-4',
'emp_length:5-6',
'emp_length:7-9',
'emp_length:10',
'mths_since_issue_d:<38',
'mths_since_issue_d:38-39',
'mths_since_issue_d:40-41',
'mths_since_issue_d:42-48',
'mths_since_issue_d:49-52',
'mths_since_issue_d:53-64',
'mths_since_issue_d:65-84',
'mths_since_issue_d:>84',
'int_rate:<9.548',
'int_rate:9.548-12.025',
'int_rate:12.025-15.74',
'int_rate:15.74-20.281',
'int_rate:>20.281',
'mths_since_earliest_cr_line:<140',
'mths_since_earliest_cr_line:141-164',
'mths_since_earliest_cr_line:165-247',
'mths_since_earliest_cr_line:248-270',
'mths_since_earliest_cr_line:271-352',
'mths_since_earliest_cr_line:>352',
'delinq_2yrs:0',
'delinq_2yrs:1-3',
'delinq_2yrs:>=4',
'inq_last_6mths:0',
'inq_last_6mths:1-2',
'inq_last_6mths:3-6',
'inq_last_6mths:>6',
'open_acc:0',
'open_acc:1-3',
'open_acc:4-12',
'open_acc:13-17',
'open_acc:18-22',
'open_acc:23-25',
'open_acc:26-30',
'open_acc:>=31',
'pub_rec:0-2',
'pub_rec:3-4',
'pub_rec:>=5',
'total_acc:<=27',
'total_acc:28-51',
'total_acc:>=52',
'acc_now_delinq:0',
'acc_now_delinq:>=1',
'total_rev_hi_lim:<=5K',
'total_rev_hi_lim:5K-10K',
'total_rev_hi_lim:10K-20K',
'total_rev_hi_lim:20K-30K',
'total_rev_hi_lim:30K-40K',
'total_rev_hi_lim:40K-55K',
'total_rev_hi_lim:55K-95K',
'total_rev_hi_lim:>95K',
'annual_inc:<20K',
'annual_inc:20K-30K',
'annual_inc:30K-40K',
'annual_inc:40K-50K',
'annual_inc:50K-60K',
'annual_inc:60K-70K',
'annual_inc:70K-80K',
'annual_inc:80K-90K',
'annual_inc:90K-100K',
'annual_inc:100K-120K',
'annual_inc:120K-140K',
'annual_inc:>140K',
'dti:<=1.4',
'dti:1.4-3.5',
'dti:3.5-7.7',
'dti:7.7-10.5',
'dti:10.5-16.1',
'dti:16.1-20.3',
'dti:20.3-21.7',
'dti:21.7-22.4',
'dti:22.4-35',
'dti:>35',
'mths_since_last_delinq:Missing',
'mths_since_last_delinq:0-3',
'mths_since_last_delinq:4-30',
'mths_since_last_delinq:31-56',
'mths_since_last_delinq:>=57',
'mths_since_last_record:Missing',
'mths_since_last_record:0-2',
'mths_since_last_record:3-20',
'mths_since_last_record:21-31',
'mths_since_last_record:32-80',
'mths_since_last_record:81-86',
'mths_since_last_record:>86',
]]

In [117]:
# Here we store the names of the reference category dummy variables in a list.
ref_categories = ['grade:G',
'home_ownership:RENT_OTHER_NONE_ANY',
'addr_state:ND_NE_IA_NV_FL_HI_AL',
'verification_status:Verified',
'purpose:educ__sm_b__wedd__ren_en__mov__house',
'initial_list_status:f',
'term:60',
'emp_length:0',
'mths_since_issue_d:>84',
'int_rate:>20.281',
'mths_since_earliest_cr_line:<140',
'delinq_2yrs:>=4',
'inq_last_6mths:>6',
'open_acc:0',
'pub_rec:0-2',
'total_acc:<=27',
'acc_now_delinq:0',
'total_rev_hi_lim:<=5K',
'annual_inc:<20K',
'dti:>35',
'mths_since_last_delinq:0-3',
'mths_since_last_record:0-2']

In [118]:
inputs_train = inputs_train_with_ref_cat.drop(ref_categories, axis = 1)
# From the dataframe with input variables, we drop the variables with variable names in the list with reference categories. 
inputs_train.head()

grade:A  grade:B  grade:C  grade:D  grade:E  grade:F  \
427211        1        0        0        0        0        0   
206088        0        0        1        0        0        0   
136020        1        0        0        0        0        0   
412305        0        0        0        1        0        0   
36159         0        0        1        0        0        0   

        home_ownership:OWN  home_ownership:MORTGAGE  addr_state:NM_VA  \
427211                   0                        1                 0   
206088                   0                        1                 0   
136020                   0                        1                 0   
412305                   0                        0                 0   
36159                    0                        1                 0   

        addr_state:NY  ...  mths_since_last_delinq:Missing  \
427211              0  ...                               1   
206088              0  ...                               0   
136020              0  ...                               0   
412305              0  ...                               0   
36159               0  ...                               1   

        mths_since_last_delinq:4-30  mths_since_last_delinq:31-56  \
427211                            0                             0   
206088                            1                             0   
136020                            0                             1   
412305                            1                             0   
36159                             0                             0   

        mths_since_last_delinq:>=57  mths_since_last_record:Missing  \
427211                            0                               1   
206088                            0                               1   
136020                            0                               1   
412305                            0                               1   
36159                             0                               1   

        mths_since_last_record:3-20  mths_since_last_record:21-31  \
427211                            0                             0   
206088                            0                             0   
136020                            0                             0   
412305                            0                             0   
36159                             0                             0   

        mths_since_last_record:32-80  mths_since_last_record:81-86  \
427211                             0                             0   
206088                             0                             0   
136020                             0                             0   
412305                             0                             0   
36159                              0                             0   

        mths_since_last_record:>86  
427211                           0  
206088                           0  
136020                           0  
412305                           0  
36159                            0  

[5 rows x 104 columns]

# PD Model Estimation

## Logistic Regression

In [121]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

In [122]:
reg = LogisticRegression()

In [123]:
pd.options.display.max_rows = None

In [124]:
reg.fit(inputs_train, loan_data_targets_train)

C:\Users\LENOVO\AppData\Roaming\Python\Python310\site-packages\sklearn\utils\validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression()

In [125]:
reg.intercept_

array([-0.44965256])

In [126]:
reg.coef_

array([[ 0.95942412,  0.75419603,  0.57382612,  0.39296023,  0.23858349,
         0.06514085,  0.09071207,  0.10272728,  0.0434879 ,  0.0700851 ,
         0.06634504,  0.070449  ,  0.08742161,  0.14728105,  0.11435442,
         0.19645364,  0.24278441,  0.23403271,  0.27507017,  0.3358681 ,
         0.49049264,  0.08290307, -0.01561263,  0.27313228,  0.17101618,
         0.17684615,  0.2297484 ,  0.0511012 ,  0.07122902,  0.10982552,
         0.13460411,  0.10251324,  0.06878645,  0.12898412,  1.12592344,
         0.93341095,  0.83522191,  0.6408304 ,  0.48088444,  0.23577322,
        -0.0161227 ,  0.97805026,  0.62682538,  0.36322798,  0.15220974,
         0.07708757,  0.05876805,  0.10545672,  0.14241762,  0.14888828,
        -0.03435282, -0.06297439,  0.28330283,  0.14081337, -0.06773606,
         0.06935818, -0.04301044, -0.06616424, -0.09375449, -0.07705438,
        -0.05423848, -0.10916809,  0.13551236,  0.09324756, -0.02115753,
         0.03160623,  0.15998118,  0.04276516,  0.0

In [127]:
feature_name = inputs_train.columns.values

In [128]:
summary_table = pd.DataFrame(columns = ['Feature Name'], data = feature_name)

#Transposing ensures that coefficients are placed in rows corresponding to feature names.
summary_table['Coefficients'] = np.transpose(reg.coef_) 

# Next we will adjust the index of the DataFrame so that the row indices start at 1 instead of 0.
summary_table.index = summary_table.index +1   

#This line adds a new row at index 0 with the Intercept value
summary_table.loc[0] = ['Intercept', reg.intercept_[0]]   

#This line sorts the DataFrame by its index in ascending order.
summary_table = summary_table.sort_index()  

summary_table

Feature Name  Coefficients
0                              Intercept     -0.449653
1                                grade:A      0.959424
2                                grade:B      0.754196
3                                grade:C      0.573826
4                                grade:D      0.392960
5                                grade:E      0.238583
6                                grade:F      0.065141
7                     home_ownership:OWN      0.090712
8                home_ownership:MORTGAGE      0.102727
9                       addr_state:NM_VA      0.043488
10                         addr_state:NY      0.070085
11          addr_state:OK_TN_MO_LA_MD_NC      0.066345
12                         addr_state:CA      0.070449
13                addr_state:UT_KY_AZ_NJ      0.087422
14             addr_state:AR_MI_PA_OH_MN      0.147281
15             addr_state:RI_MA_DE_SD_IN      0.114354
16                   addr_state:GA_WA_OR      0.196454
17                      addr_state:WI_MT      0.242784
18                         addr_state:TX      0.234033
19                      addr_state:IL_CT      0.275070
20          addr_state:KS_SC_CO_VT_AK_MS      0.335868
21          addr_state:WV_NH_WY_DC_ME_ID      0.490493
22      verification_status:Not Verified      0.082903
23   verification_status:Source Verified     -0.015613
24                   purpose:credit_card      0.273132
25            purpose:debt_consolidation      0.171016
26            purpose:oth__med__vacation      0.176846
27   purpose:major_purch__car__home_impr      0.229748
28                 initial_list_status:w      0.051101
29                               term:36      0.071229
30                          emp_length:1      0.109826
31                        emp_length:2-4      0.134604
32                        emp_length:5-6      0.102513
33                        emp_length:7-9      0.068786
34                         emp_length:10      0.128984
35                mths_since_issue_d:<38      1.125923
36              mths_since_issue_d:38-39      0.933411
37              mths_since_issue_d:40-41      0.835222
38              mths_since_issue_d:42-48      0.640830
39              mths_since_issue_d:49-52      0.480884
40              mths_since_issue_d:53-64      0.235773
41              mths_since_issue_d:65-84     -0.016123
42                       int_rate:<9.548      0.978050
43                 int_rate:9.548-12.025      0.626825
44                 int_rate:12.025-15.74      0.363228
45                 int_rate:15.74-20.281      0.152210
46   mths_since_earliest_cr_line:141-164      0.077088
47   mths_since_earliest_cr_line:165-247      0.058768
48   mths_since_earliest_cr_line:248-270      0.105457
49   mths_since_earliest_cr_line:271-352      0.142418
50      mths_since_earliest_cr_line:>352      0.148888
51                         delinq_2yrs:0     -0.034353
52                       delinq_2yrs:1-3     -0.062974
53                      inq_last_6mths:0      0.283303
54                    inq_last_6mths:1-2      0.140813
55                    inq_last_6mths:3-6     -0.067736
56                          open_acc:1-3      0.069358
57                         open_acc:4-12     -0.043010
58                        open_acc:13-17     -0.066164
59                        open_acc:18-22     -0.093754
60                        open_acc:23-25     -0.077054
61                        open_acc:26-30     -0.054238
62                         open_acc:>=31     -0.109168
63                           pub_rec:3-4      0.135512
64                           pub_rec:>=5      0.093248
65                       total_acc:28-51     -0.021158
66                        total_acc:>=52      0.031606
67                    acc_now_delinq:>=1      0.159981
68               total_rev_hi_lim:5K-10K      0.042765
69              total_rev_hi_lim:10K-20K      0.008320
70              total_rev_hi_lim:20K-30K      0.002484
71              total_rev_hi_lim:30K-40K      0.020059
72           

## Building a Logistic Regression Model with P-values

In [130]:
from sklearn import linear_model
import scipy.stats as stat

Unfortunately, Logistic Regression does not have an inbuilt way to calculate p-values.<br>
So, we will need to use linear model here for getting our p-values

In [132]:
class LogisticRegression_with_p_values:
    def __init__(self, *args, **kwargs):
        self.model = linear_model.LogisticRegression(*args, **kwargs)

    #Next we overwrite .fit() with a function that calculate p-values too
    def fit(self, X,y):
        self.model.fit(X,y)
        denom = (2.0*(1.0+np.cosh(self.model.decision_function(X))))
        denom = np.tile(denom,(X.shape[1],1)).T
        F_ij = np.dot((X/denom).T,X)
        Cramer_Rao = np.linalg.inv(F_ij)
        sigma_estimates = np.sqrt(np.diagonal(Cramer_Rao))
        z_scores = self.model.coef_[0]/sigma_estimates
        p_values = [stat.norm.sf(abs(x))*2 for x in z_scores]
        self.coef_ = self.model.coef_
        self.intercept_ = self.model.intercept_
        self.p_values = p_values

In [133]:
reg = LogisticRegression_with_p_values()

In [134]:
reg.fit(inputs_train, loan_data_targets_train)

C:\Users\LENOVO\AppData\Roaming\Python\Python310\site-packages\sklearn\utils\validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [135]:
summary_table = pd.DataFrame(columns = ['Feature Name'], data = feature_name)

#Transposing ensures that coefficients are placed in rows corresponding to feature names.
summary_table['Coefficients'] = np.transpose(reg.coef_) 

# Next we will adjust the index of the DataFrame so that the row indices start at 1 instead of 0.
summary_table.index = summary_table.index +1   

#This line adds a new row at index 0 with the Intercept value
summary_table.loc[0] = ['Intercept', reg.intercept_[0]]   

#This line sorts the DataFrame by its index in ascending order.
summary_table = summary_table.sort_index()  

summary_table

Feature Name  Coefficients
0                              Intercept     -0.449653
1                                grade:A      0.959424
2                                grade:B      0.754196
3                                grade:C      0.573826
4                                grade:D      0.392960
5                                grade:E      0.238583
6                                grade:F      0.065141
7                     home_ownership:OWN      0.090712
8                home_ownership:MORTGAGE      0.102727
9                       addr_state:NM_VA      0.043488
10                         addr_state:NY      0.070085
11          addr_state:OK_TN_MO_LA_MD_NC      0.066345
12                         addr_state:CA      0.070449
13                addr_state:UT_KY_AZ_NJ      0.087422
14             addr_state:AR_MI_PA_OH_MN      0.147281
15             addr_state:RI_MA_DE_SD_IN      0.114354
16                   addr_state:GA_WA_OR      0.196454
17                      addr_state:WI_MT      0.242784
18                         addr_state:TX      0.234033
19                      addr_state:IL_CT      0.275070
20          addr_state:KS_SC_CO_VT_AK_MS      0.335868
21          addr_state:WV_NH_WY_DC_ME_ID      0.490493
22      verification_status:Not Verified      0.082903
23   verification_status:Source Verified     -0.015613
24                   purpose:credit_card      0.273132
25            purpose:debt_consolidation      0.171016
26            purpose:oth__med__vacation      0.176846
27   purpose:major_purch__car__home_impr      0.229748
28                 initial_list_status:w      0.051101
29                               term:36      0.071229
30                          emp_length:1      0.109826
31                        emp_length:2-4      0.134604
32                        emp_length:5-6      0.102513
33                        emp_length:7-9      0.068786
34                         emp_length:10      0.128984
35                mths_since_issue_d:<38      1.125923
36              mths_since_issue_d:38-39      0.933411
37              mths_since_issue_d:40-41      0.835222
38              mths_since_issue_d:42-48      0.640830
39              mths_since_issue_d:49-52      0.480884
40              mths_since_issue_d:53-64      0.235773
41              mths_since_issue_d:65-84     -0.016123
42                       int_rate:<9.548      0.978050
43                 int_rate:9.548-12.025      0.626825
44                 int_rate:12.025-15.74      0.363228
45                 int_rate:15.74-20.281      0.152210
46   mths_since_earliest_cr_line:141-164      0.077088
47   mths_since_earliest_cr_line:165-247      0.058768
48   mths_since_earliest_cr_line:248-270      0.105457
49   mths_since_earliest_cr_line:271-352      0.142418
50      mths_since_earliest_cr_line:>352      0.148888
51                         delinq_2yrs:0     -0.034353
52                       delinq_2yrs:1-3     -0.062974
53                      inq_last_6mths:0      0.283303
54                    inq_last_6mths:1-2      0.140813
55                    inq_last_6mths:3-6     -0.067736
56                          open_acc:1-3      0.069358
57                         open_acc:4-12     -0.043010
58                        open_acc:13-17     -0.066164
59                        open_acc:18-22     -0.093754
60                        open_acc:23-25     -0.077054
61                        open_acc:26-30     -0.054238
62                         open_acc:>=31     -0.109168
63                           pub_rec:3-4      0.135512
64                           pub_rec:>=5      0.093248
65                       total_acc:28-51     -0.021158
66                        total_acc:>=52      0.031606
67                    acc_now_delinq:>=1      0.159981
68               total_rev_hi_lim:5K-10K      0.042765
69              total_rev_hi_lim:10K-20K      0.008320
70              total_rev_hi_lim:20K-30K      0.002484
71              total_rev_hi_lim:30K-40K      0.020059
72           

In [136]:
p_values = reg.p_values

In [137]:
p_values = np.append(np.nan, np.array(p_values))

In [138]:
summary_table['p_values'] = p_values

In [139]:
summary_table

Feature Name  Coefficients       p_values
0                              Intercept     -0.449653            NaN
1                                grade:A      0.959424   1.123162e-25
2                                grade:B      0.754196   2.546667e-34
3                                grade:C      0.573826   1.414468e-23
4                                grade:D      0.392960   5.679713e-13
5                                grade:E      0.238583   1.032942e-06
6                                grade:F      0.065141   2.004100e-01
7                     home_ownership:OWN      0.090712   7.089808e-06
8                home_ownership:MORTGAGE      0.102727   5.937190e-16
9                       addr_state:NM_VA      0.043488   1.745103e-01
10                         addr_state:NY      0.070085   3.885191e-03
11          addr_state:OK_TN_MO_LA_MD_NC      0.066345   4.198037e-03
12                         addr_state:CA      0.070449   8.937483e-04
13                addr_state:UT_KY_AZ_NJ      0.087422   4.535210e-04
14             addr_state:AR_MI_PA_OH_MN      0.147281   6.664833e-11
15             addr_state:RI_MA_DE_SD_IN      0.114354   1.304977e-04
16                   addr_state:GA_WA_OR      0.196454   1.895935e-13
17                      addr_state:WI_MT      0.242784   3.218526e-07
18                         addr_state:TX      0.234033   2.243400e-19
19                      addr_state:IL_CT      0.275070   5.431586e-21
20          addr_state:KS_SC_CO_VT_AK_MS      0.335868   1.677965e-27
21          addr_state:WV_NH_WY_DC_ME_ID      0.490493   8.998227e-21
22      verification_status:Not Verified      0.082903   1.707548e-08
23   verification_status:Source Verified     -0.015613   2.466480e-01
24                   purpose:credit_card      0.273132   1.406704e-21
25            purpose:debt_consolidation      0.171016   8.762676e-11
26            purpose:oth__med__vacation      0.176846   1.731071e-08
27   purpose:major_purch__car__home_impr      0.229748   3.051693e-13
28                 initial_list_status:w      0.051101   9.118707e-05
29                               term:36      0.071229   5.116585e-07
30                          emp_length:1      0.109826   2.345302e-05
31                        emp_length:2-4      0.134604   1.001288e-12
32                        emp_length:5-6      0.102513   1.684249e-06
33                        emp_length:7-9      0.068786   9.385066e-04
34                         emp_length:10      0.128984   2.563952e-12
35                mths_since_issue_d:<38      1.125923  9.825824e-190
36              mths_since_issue_d:38-39      0.933411  2.222756e-162
37              mths_since_issue_d:40-41      0.835222  9.785001e-137
38              mths_since_issue_d:42-48      0.640830  9.331976e-103
39              mths_since_issue_d:49-52      0.480884   7.625745e-52
40              mths_since_issue_d:53-64      0.235773   2.898279e-15
41              mths_since_issue_d:65-84     -0.016123   5.789051e-01
42                       int_rate:<9.548      0.978050   3.372954e-37
43                 int_rate:9.548-12.025      0.626825   3.419023e-44
44                 int_rate:12.025-15.74      0.363228   9.519253e-23
45                 int_rate:15.74-20.281      0.152210   1.842180e-07
46   mths_since_earliest_cr_line:141-164      0.077088   2.002972e-03
47   mths_since_earliest_cr_line:165-247      0.058768   3.817462e-03
48   mths_since_earliest_cr_line:248-270      0.105457   7.283339e-05
49   mths_since_earliest_cr_line:271-352      0.142418   2.410748e-09
50      mths_since_earliest_cr_line:>352      0.148888   3.558723e-08
51                         delinq_2yrs:0     -0.034353   5.585025e-01
52                       delinq_2yrs:1-3     -0.062974   2.561635e-01
53                      inq_last_6mths:0      0.283303   2.107527e-02
54                    inq_last_6mths:1-2      0.140813   2.510080e-01
55                    inq_last_6mths:3-6     -0.067736   5.822204e-01
56                          open_acc:1

In [140]:
summary_table['p_values'] = summary_table['p_values'].apply(lambda x: f'{x:.10f}')
summary_table

Feature Name  Coefficients      p_values
0                              Intercept     -0.449653           nan
1                                grade:A      0.959424  0.0000000000
2                                grade:B      0.754196  0.0000000000
3                                grade:C      0.573826  0.0000000000
4                                grade:D      0.392960  0.0000000000
5                                grade:E      0.238583  0.0000010329
6                                grade:F      0.065141  0.2004099842
7                     home_ownership:OWN      0.090712  0.0000070898
8                home_ownership:MORTGAGE      0.102727  0.0000000000
9                       addr_state:NM_VA      0.043488  0.1745103141
10                         addr_state:NY      0.070085  0.0038851913
11          addr_state:OK_TN_MO_LA_MD_NC      0.066345  0.0041980375
12                         addr_state:CA      0.070449  0.0008937483
13                addr_state:UT_KY_AZ_NJ      0.087422  0.0004535210
14             addr_state:AR_MI_PA_OH_MN      0.147281  0.0000000001
15             addr_state:RI_MA_DE_SD_IN      0.114354  0.0001304977
16                   addr_state:GA_WA_OR      0.196454  0.0000000000
17                      addr_state:WI_MT      0.242784  0.0000003219
18                         addr_state:TX      0.234033  0.0000000000
19                      addr_state:IL_CT      0.275070  0.0000000000
20          addr_state:KS_SC_CO_VT_AK_MS      0.335868  0.0000000000
21          addr_state:WV_NH_WY_DC_ME_ID      0.490493  0.0000000000
22      verification_status:Not Verified      0.082903  0.0000000171
23   verification_status:Source Verified     -0.015613  0.2466479727
24                   purpose:credit_card      0.273132  0.0000000000
25            purpose:debt_consolidation      0.171016  0.0000000001
26            purpose:oth__med__vacation      0.176846  0.0000000173
27   purpose:major_purch__car__home_impr      0.229748  0.0000000000
28                 initial_list_status:w      0.051101  0.0000911871
29                               term:36      0.071229  0.0000005117
30                          emp_length:1      0.109826  0.0000234530
31                        emp_length:2-4      0.134604  0.0000000000
32                        emp_length:5-6      0.102513  0.0000016842
33                        emp_length:7-9      0.068786  0.0009385066
34                         emp_length:10      0.128984  0.0000000000
35                mths_since_issue_d:<38      1.125923  0.0000000000
36              mths_since_issue_d:38-39      0.933411  0.0000000000
37              mths_since_issue_d:40-41      0.835222  0.0000000000
38              mths_since_issue_d:42-48      0.640830  0.0000000000
39              mths_since_issue_d:49-52      0.480884  0.0000000000
40              mths_since_issue_d:53-64      0.235773  0.0000000000
41              mths_since_issue_d:65-84     -0.016123  0.5789050527
42                       int_rate:<9.548      0.978050  0.0000000000
43                 int_rate:9.548-12.025      0.626825  0.0000000000
44                 int_rate:12.025-15.74      0.363228  0.0000000000
45                 int_rate:15.74-20.281      0.152210  0.0000001842
46   mths_since_earliest_cr_line:141-164      0.077088  0.0020029717
47   mths_since_earliest_cr_line:165-247      0.058768  0.0038174625
48   mths_since_earliest_cr_line:248-270      0.105457  0.0000728334
49   mths_since_earliest_cr_line:271-352      0.142418  0.0000000024
50      mths_since_earliest_cr_line:>352      0.148888  0.0000000356
51                         delinq_2yrs:0     -0.034353  0.5585024971
52                       delinq_2yrs:1-3     -0.062974  0.2561635229
53                      inq_last_6mths:0      0.283303  0.0210752717
54                    inq_last_6mths:1-2      0.140813  0.2510079561
55                    inq_last_6mths:3-6     -0.067736  0.5822204341
56                          open_acc:1-3      0.069358  0.6895564231
57                        

P-value plays an important role in our model. It tells us which feature is statistically significant. <br>
If the p-value is less than 0.05 or 5%, we will consider the feature to be statistically significant. <br>

We are going to remove some features, the coefficients for all or almost all of the dummy variables for which,
are not tatistically significant.

We do that by specifying another list of dummy variables as reference categories, and a list of variables to remove.<br>
Then, we are going to drop the two datasets from the original list of dummy variables.


In [142]:
# Variables
inputs_train_with_ref_cat = loan_data_inputs_train.loc[: , ['grade:A',
'grade:B',
'grade:C',
'grade:D',
'grade:E',
'grade:F',
'grade:G',
'home_ownership:RENT_OTHER_NONE_ANY',
'home_ownership:OWN',
'home_ownership:MORTGAGE',
'addr_state:ND_NE_IA_NV_FL_HI_AL',
'addr_state:NM_VA',
'addr_state:NY',
'addr_state:OK_TN_MO_LA_MD_NC',
'addr_state:CA',
'addr_state:UT_KY_AZ_NJ',
'addr_state:AR_MI_PA_OH_MN',
'addr_state:RI_MA_DE_SD_IN',
'addr_state:GA_WA_OR',
'addr_state:WI_MT',
'addr_state:TX',
'addr_state:IL_CT',
'addr_state:KS_SC_CO_VT_AK_MS',
'addr_state:WV_NH_WY_DC_ME_ID',
'verification_status:Not Verified',
'verification_status:Source Verified',
'verification_status:Verified',
'purpose:educ__sm_b__wedd__ren_en__mov__house',
'purpose:credit_card',
'purpose:debt_consolidation',
'purpose:oth__med__vacation',
'purpose:major_purch__car__home_impr',
'initial_list_status:f',
'initial_list_status:w',
'term:36',
'term:60',
'emp_length:0',
'emp_length:1',
'emp_length:2-4',
'emp_length:5-6',
'emp_length:7-9',
'emp_length:10',
'mths_since_issue_d:<38',
'mths_since_issue_d:38-39',
'mths_since_issue_d:40-41',
'mths_since_issue_d:42-48',
'mths_since_issue_d:49-52',
'mths_since_issue_d:53-64',
'mths_since_issue_d:65-84',
'mths_since_issue_d:>84',
'int_rate:<9.548',
'int_rate:9.548-12.025',
'int_rate:12.025-15.74',
'int_rate:15.74-20.281',
'int_rate:>20.281',
'mths_since_earliest_cr_line:<140',
'mths_since_earliest_cr_line:141-164',
'mths_since_earliest_cr_line:165-247',
'mths_since_earliest_cr_line:248-270',
'mths_since_earliest_cr_line:271-352',
'mths_since_earliest_cr_line:>352',
'inq_last_6mths:0',
'inq_last_6mths:1-2',
'inq_last_6mths:3-6',
'inq_last_6mths:>6',
'acc_now_delinq:0',
'acc_now_delinq:>=1',
'annual_inc:<20K',
'annual_inc:20K-30K',
'annual_inc:30K-40K',
'annual_inc:40K-50K',
'annual_inc:50K-60K',
'annual_inc:60K-70K',
'annual_inc:70K-80K',
'annual_inc:80K-90K',
'annual_inc:90K-100K',
'annual_inc:100K-120K',
'annual_inc:120K-140K',
'annual_inc:>140K',
'dti:<=1.4',
'dti:1.4-3.5',
'dti:3.5-7.7',
'dti:7.7-10.5',
'dti:10.5-16.1',
'dti:16.1-20.3',
'dti:20.3-21.7',
'dti:21.7-22.4',
'dti:22.4-35',
'dti:>35',
'mths_since_last_delinq:Missing',
'mths_since_last_delinq:0-3',
'mths_since_last_delinq:4-30',
'mths_since_last_delinq:31-56',
'mths_since_last_delinq:>=57',
'mths_since_last_record:Missing',
'mths_since_last_record:0-2',
'mths_since_last_record:3-20',
'mths_since_last_record:21-31',
'mths_since_last_record:32-80',
'mths_since_last_record:81-86',
'mths_since_last_record:>86',
]]

In [143]:
ref_categories = ['grade:G',
'home_ownership:RENT_OTHER_NONE_ANY',
'addr_state:ND_NE_IA_NV_FL_HI_AL',
'verification_status:Verified',
'purpose:educ__sm_b__wedd__ren_en__mov__house',
'initial_list_status:f',
'term:60',
'emp_length:0',
'mths_since_issue_d:>84',
'int_rate:>20.281',
'mths_since_earliest_cr_line:<140',
'inq_last_6mths:>6',
'acc_now_delinq:0',
'annual_inc:<20K',
'dti:>35',
'mths_since_last_delinq:0-3',
'mths_since_last_record:0-2']

In [144]:
inputs_train = inputs_train_with_ref_cat.drop(ref_categories, axis = 1)
inputs_train.head()

grade:A  grade:B  grade:C  grade:D  grade:E  grade:F  \
427211        1        0        0        0        0        0   
206088        0        0        1        0        0        0   
136020        1        0        0        0        0        0   
412305        0        0        0        1        0        0   
36159         0        0        1        0        0        0   

        home_ownership:OWN  home_ownership:MORTGAGE  addr_state:NM_VA  \
427211                   0                        1                 0   
206088                   0                        1                 0   
136020                   0                        1                 0   
412305                   0                        0                 0   
36159                    0                        1                 0   

        addr_state:NY  ...  mths_since_last_delinq:Missing  \
427211              0  ...                               1   
206088              0  ...                               0   
136020              0  ...                               0   
412305              0  ...                               0   
36159               0  ...                               1   

        mths_since_last_delinq:4-30  mths_since_last_delinq:31-56  \
427211                            0                             0   
206088                            1                             0   
136020                            0                             1   
412305                            1                             0   
36159                             0                             0   

        mths_since_last_delinq:>=57  mths_since_last_record:Missing  \
427211                            0                               1   
206088                            0                               1   
136020                            0                               1   
412305                            0                               1   
36159                             0                               1   

        mths_since_last_record:3-20  mths_since_last_record:21-31  \
427211                            0                             0   
206088                            0                             0   
136020                            0                             0   
412305                            0                             0   
36159                             0                             0   

        mths_since_last_record:32-80  mths_since_last_record:81-86  \
427211                             0                             0   
206088                             0                             0   
136020                             0                             0   
412305                             0                             0   
36159                              0                             0   

        mths_since_last_record:>86  
427211                           0  
206088                           0  
136020                           0  
412305                           0  
36159                            0  

[5 rows x 84 columns]

In [145]:
# We once again, run our model!
reg2 = LogisticRegression_with_p_values()
reg2.fit(inputs_train, loan_data_targets_train)

C:\Users\LENOVO\AppData\Roaming\Python\Python310\site-packages\sklearn\utils\validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [146]:
feature_name = inputs_train.columns.values
# Same as above.
summary_table = pd.DataFrame(columns = ['Feature name'], data = feature_name)
summary_table['Coefficients'] = np.transpose(reg2.coef_)
summary_table.index = summary_table.index + 1
summary_table.loc[0] = ['Intercept', reg2.intercept_[0]]
summary_table = summary_table.sort_index()
summary_table

Feature name  Coefficients
0                             Intercept     -0.463330
1                               grade:A      0.942864
2                               grade:B      0.730167
3                               grade:C      0.547289
4                               grade:D      0.372100
5                               grade:E      0.218311
6                               grade:F      0.049030
7                    home_ownership:OWN      0.096190
8               home_ownership:MORTGAGE      0.104737
9                      addr_state:NM_VA      0.044846
10                        addr_state:NY      0.069922
11         addr_state:OK_TN_MO_LA_MD_NC      0.066962
12                        addr_state:CA      0.073659
13               addr_state:UT_KY_AZ_NJ      0.088810
14            addr_state:AR_MI_PA_OH_MN      0.145877
15            addr_state:RI_MA_DE_SD_IN      0.113823
16                  addr_state:GA_WA_OR      0.196644
17                     addr_state:WI_MT      0.243645
18                        addr_state:TX      0.232673
19                     addr_state:IL_CT      0.272130
20         addr_state:KS_SC_CO_VT_AK_MS      0.334345
21         addr_state:WV_NH_WY_DC_ME_ID      0.501561
22     verification_status:Not Verified      0.080365
23  verification_status:Source Verified     -0.013023
24                  purpose:credit_card      0.283290
25           purpose:debt_consolidation      0.177846
26           purpose:oth__med__vacation      0.188264
27  purpose:major_purch__car__home_impr      0.242061
28                initial_list_status:w      0.049319
29                              term:36      0.080711
30                         emp_length:1      0.109178
31                       emp_length:2-4      0.133497
32                       emp_length:5-6      0.100144
33                       emp_length:7-9      0.067044
34                        emp_length:10      0.129904
35               mths_since_issue_d:<38      1.133319
36             mths_since_issue_d:38-39      0.943318
37             mths_since_issue_d:40-41      0.843239
38             mths_since_issue_d:42-48      0.651742
39             mths_since_issue_d:49-52      0.487715
40             mths_since_issue_d:53-64      0.242539
41             mths_since_issue_d:65-84     -0.008368
42                      int_rate:<9.548      0.965410
43                int_rate:9.548-12.025      0.627531
44                int_rate:12.025-15.74      0.362496
45                int_rate:15.74-20.281      0.150002
46  mths_since_earliest_cr_line:141-164      0.076871
47  mths_since_earliest_cr_line:165-247      0.057056
48  mths_since_earliest_cr_line:248-270      0.105700
49  mths_since_earliest_cr_line:271-352      0.145977
50     mths_since_earliest_cr_line:>352      0.153023
51                     inq_last_6mths:0      0.265943
52                   inq_last_6mths:1-2      0.121518
53                   inq_last_6mths:3-6     -0.086156
54                   acc_now_delinq:>=1      0.145208
55                   annual_inc:20K-30K     -0.180762
56                   annual_inc:30K-40K     -0.117121
57                   annual_inc:40K-50K     -0.022238
58                   annual_inc:50K-60K      0.042513
59                   annual_inc:60K-70K      0.121089
60                   annual_inc:70K-80K      0.192124
61                   annual_inc:80K-90K      0.268270
62                  annual_inc:90K-100K      0.279159
63                 annual_inc:100K-120K      0.359531
64                 annual_inc:120K-140K      0.480280
65                     annual_inc:>140K      0.386156
66                            dti:<=1.4      0.057756
67                          dti:1.4-3.5      0.167917
68                          dti:3.5-7.7      0.201357
69                         dti:7.7-10.5      0.134914
70                        dti:10.5-16.1      0.054082
71                        dti:16.1-20.3     -0.047825
72                        dti:20.3-21.7     -0.065848
73                        dti:2

In [147]:
# We add the 'p_values' here, just as we did before.
p_values = reg2.p_values
p_values = np.append(np.nan,np.array(p_values))
summary_table['p_values'] = p_values
summary_table['p_values'] = summary_table['p_values'].apply(lambda x: f'{x:.10f}')
summary_table
# Here we get the results for our final PD model.

Feature name  Coefficients      p_values
0                             Intercept     -0.463330           nan
1                               grade:A      0.942864  0.0000000000
2                               grade:B      0.730167  0.0000000000
3                               grade:C      0.547289  0.0000000000
4                               grade:D      0.372100  0.0000000000
5                               grade:E      0.218311  0.0000059110
6                               grade:F      0.049030  0.3272023206
7                    home_ownership:OWN      0.096190  0.0000018698
8               home_ownership:MORTGAGE      0.104737  0.0000000000
9                      addr_state:NM_VA      0.044846  0.1609073870
10                        addr_state:NY      0.069922  0.0038713327
11         addr_state:OK_TN_MO_LA_MD_NC      0.066962  0.0037719313
12                        addr_state:CA      0.073659  0.0005001196
13               addr_state:UT_KY_AZ_NJ      0.088810  0.0003588703
14            addr_state:AR_MI_PA_OH_MN      0.145877  0.0000000001
15            addr_state:RI_MA_DE_SD_IN      0.113823  0.0001364704
16                  addr_state:GA_WA_OR      0.196644  0.0000000000
17                     addr_state:WI_MT      0.243645  0.0000002878
18                        addr_state:TX      0.232673  0.0000000000
19                     addr_state:IL_CT      0.272130  0.0000000000
20         addr_state:KS_SC_CO_VT_AK_MS      0.334345  0.0000000000
21         addr_state:WV_NH_WY_DC_ME_ID      0.501561  0.0000000000
22     verification_status:Not Verified      0.080365  0.0000000365
23  verification_status:Source Verified     -0.013023  0.3323415107
24                  purpose:credit_card      0.283290  0.0000000000
25           purpose:debt_consolidation      0.177846  0.0000000000
26           purpose:oth__med__vacation      0.188264  0.0000000016
27  purpose:major_purch__car__home_impr      0.242061  0.0000000000
28                initial_list_status:w      0.049319  0.0001584992
29                              term:36      0.080711  0.0000000078
30                         emp_length:1      0.109178  0.0000257849
31                       emp_length:2-4      0.133497  0.0000000000
32                       emp_length:5-6      0.100144  0.0000028517
33                       emp_length:7-9      0.067044  0.0012542208
34                        emp_length:10      0.129904  0.0000000000
35               mths_since_issue_d:<38      1.133319  0.0000000000
36             mths_since_issue_d:38-39      0.943318  0.0000000000
37             mths_since_issue_d:40-41      0.843239  0.0000000000
38             mths_since_issue_d:42-48      0.651742  0.0000000000
39             mths_since_issue_d:49-52      0.487715  0.0000000000
40             mths_since_issue_d:53-64      0.242539  0.0000000000
41             mths_since_issue_d:65-84     -0.008368  0.7721670409
42                      int_rate:<9.548      0.965410  0.0000000000
43                int_rate:9.548-12.025      0.627531  0.0000000000
44                int_rate:12.025-15.74      0.362496  0.0000000000
45                int_rate:15.74-20.281      0.150002  0.0000002364
46  mths_since_earliest_cr_line:141-164      0.076871  0.0019843789
47  mths_since_earliest_cr_line:165-247      0.057056  0.0043986290
48  mths_since_earliest_cr_line:248-270      0.105700  0.0000555468
49  mths_since_earliest_cr_line:271-352      0.145977  0.0000000004
50     mths_since_earliest_cr_line:>352      0.153023  0.0000000076
51                     inq_last_6mths:0      0.265943  0.0042371438
52                   inq_last_6mths:1-2      0.121518  0.1907874839
53                   inq_last_6mths:3-6     -0.086156  0.3562867568
54                   acc_now_delinq:>=1      0.145208  0.1730172721
55                   annual_inc:20K-30K     -0.180762  0.0000279454
56                   annual_inc:30K-40K     -0.117121  0.0048210337
57                   annual_inc:40K-50K     -0.022238  0.5920573809
58             

#### Viola! 
<br>The final summary table contain 84 feature names and it's respective data

## Output of Sample Test

In [150]:
inputs_test_with_ref_cat = loan_data_inputs_test.loc[:,['grade:A',
'grade:B',
'grade:C',
'grade:D',
'grade:E',
'grade:F',
'grade:G',
'home_ownership:RENT_OTHER_NONE_ANY',
'home_ownership:OWN',
'home_ownership:MORTGAGE',
'addr_state:ND_NE_IA_NV_FL_HI_AL',
'addr_state:NM_VA',
'addr_state:NY',
'addr_state:OK_TN_MO_LA_MD_NC',
'addr_state:CA',
'addr_state:UT_KY_AZ_NJ',
'addr_state:AR_MI_PA_OH_MN',
'addr_state:RI_MA_DE_SD_IN',
'addr_state:GA_WA_OR',
'addr_state:WI_MT',
'addr_state:TX',
'addr_state:IL_CT',
'addr_state:KS_SC_CO_VT_AK_MS',
'addr_state:WV_NH_WY_DC_ME_ID',
'verification_status:Not Verified',
'verification_status:Source Verified',
'verification_status:Verified',
'purpose:educ__sm_b__wedd__ren_en__mov__house',
'purpose:credit_card',
'purpose:debt_consolidation',
'purpose:oth__med__vacation',
'purpose:major_purch__car__home_impr',
'initial_list_status:f',
'initial_list_status:w',
'term:36',
'term:60',
'emp_length:0',
'emp_length:1',
'emp_length:2-4',
'emp_length:5-6',
'emp_length:7-9',
'emp_length:10',
'mths_since_issue_d:<38',
'mths_since_issue_d:38-39',
'mths_since_issue_d:40-41',
'mths_since_issue_d:42-48',
'mths_since_issue_d:49-52',
'mths_since_issue_d:53-64',
'mths_since_issue_d:65-84',
'mths_since_issue_d:>84',
'int_rate:<9.548',
'int_rate:9.548-12.025',
'int_rate:12.025-15.74',
'int_rate:15.74-20.281',
'int_rate:>20.281',
'mths_since_earliest_cr_line:<140',
'mths_since_earliest_cr_line:141-164',
'mths_since_earliest_cr_line:165-247',
'mths_since_earliest_cr_line:248-270',
'mths_since_earliest_cr_line:271-352',
'mths_since_earliest_cr_line:>352',
'inq_last_6mths:0',
'inq_last_6mths:1-2',
'inq_last_6mths:3-6',
'inq_last_6mths:>6',
'acc_now_delinq:0',
'acc_now_delinq:>=1',
'annual_inc:<20K',
'annual_inc:20K-30K',
'annual_inc:30K-40K',
'annual_inc:40K-50K',
'annual_inc:50K-60K',
'annual_inc:60K-70K',
'annual_inc:70K-80K',
'annual_inc:80K-90K',
'annual_inc:90K-100K',
'annual_inc:100K-120K',
'annual_inc:120K-140K',
'annual_inc:>140K',
'dti:<=1.4',
'dti:1.4-3.5',
'dti:3.5-7.7',
'dti:7.7-10.5',
'dti:10.5-16.1',
'dti:16.1-20.3',
'dti:20.3-21.7',
'dti:21.7-22.4',
'dti:22.4-35',
'dti:>35',
'mths_since_last_delinq:Missing',
'mths_since_last_delinq:0-3',
'mths_since_last_delinq:4-30',
'mths_since_last_delinq:31-56',
'mths_since_last_delinq:>=57',
'mths_since_last_record:Missing',
'mths_since_last_record:0-2',
'mths_since_last_record:3-20',
'mths_since_last_record:21-31',
'mths_since_last_record:32-80',
'mths_since_last_record:81-86',
'mths_since_last_record:>86'
]]

In [151]:
inputs_test = inputs_test_with_ref_cat.drop(ref_categories, axis = 1)
inputs_test.head()

grade:A  grade:B  grade:C  grade:D  grade:E  grade:F  \
362514        0        0        1        0        0        0   
288564        0        0        0        0        1        0   
213591        0        0        1        0        0        0   
263083        0        0        1        0        0        0   
165001        1        0        0        0        0        0   

        home_ownership:OWN  home_ownership:MORTGAGE  addr_state:NM_VA  \
362514                   0                        1                 0   
288564                   0                        1                 0   
213591                   0                        1                 0   
263083                   0                        1                 0   
165001                   0                        1                 0   

        addr_state:NY  ...  mths_since_last_delinq:Missing  \
362514              0  ...                               1   
288564              0  ...                               0   
213591              0  ...                               0   
263083              0  ...                               1   
165001              0  ...                               0   

        mths_since_last_delinq:4-30  mths_since_last_delinq:31-56  \
362514                            0                             0   
288564                            0                             0   
213591                            0                             1   
263083                            0                             0   
165001                            0                             1   

        mths_since_last_delinq:>=57  mths_since_last_record:Missing  \
362514                            0                               1   
288564                            0                               1   
213591                            0                               1   
263083                            0                               1   
165001                            0                               1   

        mths_since_last_record:3-20  mths_since_last_record:21-31  \
362514                            0                             0   
288564                            0                             0   
213591                            0                             0   
263083                            0                             0   
165001                            0                             0   

        mths_since_last_record:32-80  mths_since_last_record:81-86  \
362514                             0                             0   
288564                             0                             0   
213591                             0                             0   
263083                             0                             0   
165001                             0                             0   

        mths_since_last_record:>86  
362514                           0  
288564                           0  
213591                           0  
263083                           0  
165001                           0  

[5 rows x 84 columns]

In [152]:
y_hat_test = reg2.model.predict(inputs_test)

In [153]:
y_hat_test

array([1, 1, 1, ..., 1, 1, 1], dtype=int64)

In [193]:
#Now we will calculate the Probability of default and non-default.

# In this case, the first value of every sub-array is the probability for the observation to belong to the first class, i.e. 0,
# and the second value is the probability for the observation to belong to the first class, i.e. 1.

y_hat_test_proba = reg2.model.predict_proba(inputs_test)
y_hat_test_proba

array([[0.07578703, 0.92421297],
       [0.14921606, 0.85078394],
       [0.11485169, 0.88514831],
       ...,
       [0.02691035, 0.97308965],
       [0.04022598, 0.95977402],
       [0.0479313 , 0.9520687 ]])

In [ ]:
# Here we take all the arrays in the array, and from each array, we take all rows, and only the element with index 1,
# that is, the second element.
y_hat_test_proba = y_hat_test_proba[: ][: , 1]

In [ ]:
y_hat_test_proba

In [ ]:
loan_data_targets_test_temp = loan_data_targets_test

In [ ]:
loan_data_targets_test_temp.reset_index(drop = True, inplace = True)

In [ ]:
df_actual_predicted_probs = pd.concat([loan_data_targets_test_temp, pd.DataFrame(y_hat_test_proba)], axis = 1)

In [ ]:
df_actual_predicted_probs.columns = ['loan_data_targets_test', 'y_hat_test_proba']

In [ ]:
df_actual_predicted_probs.index = loan_data_inputs_test.index

In [ ]:
df_actual_predicted_probs.head()

## Accuracy and the Area Under the Curve

In [ ]:
#tr= 0.5
tr = 0.9
df_actual_predicted_probs['y_hat_test'] = np.where(df_actual_predicted_probs['y_hat_test_proba']>tr,1,0)

In [ ]:
df_actual_predicted_probs.head()

In [ ]:
pd.crosstab(df_actual_predicted_probs['loan_data_targets_test'], df_actual_predicted_probs['y_hat_test'],
           rownames = ['Actual'], colnames = ['Predicted'])

In [ ]:
pd.crosstab(df_actual_predicted_probs['loan_data_targets_test'], df_actual_predicted_probs['y_hat_test'],
           rownames = ['Actual'], colnames = ['Predicted'])/df_actual_predicted_probs.shape[0]

Now, when we study the confusion matrix we can see how many are predicted to be possibly defaulting, which is very low.

This is because we have set the default threshold at 0.5. 

These numbers are completely unacceptable, because this number suggests that very few almost none, default.

As this is not true, we will need to set a higher threshold. 

Higher the threshold, better the accuracy.

Now we will update our threshold value to 0.9, i.e. tr = 0.9 and run all the cells below it. 

In [ ]:
(pd.crosstab(df_actual_predicted_probs['loan_data_targets_test'],  df_actual_predicted_probs['y_hat_test'], 
                           rownames = ['Actual'], colnames = ['Predicted']) / df_actual_predicted_probs.shape[0]).iloc[0, 0] + (pd.crosstab(df_actual_predicted_probs['loan_data_targets_test'], 
                           df_actual_predicted_probs['y_hat_test'], 
                           rownames = ['Actual'], colnames = ['Predicted']) / df_actual_predicted_probs.shape[0]).iloc[1, 1]
# Here we calculate Accuracy of the model, which is the sum of the diagonal rates.

In [ ]:
from sklearn.metrics import roc_curve, roc_auc_score

In [ ]:
roc_curve(df_actual_predicted_probs['loan_data_targets_test'], df_actual_predicted_probs['y_hat_test_proba'])

In [ ]:
fpr, tpr, thresholds = roc_curve(df_actual_predicted_probs['loan_data_targets_test'], df_actual_predicted_probs['y_hat_test_proba'])

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

In [ ]:
plt.plot(fpr, tpr)
# We plot the false positive rate along the x-axis and the true positive rate along the y-axis,
# thus plotting the ROC curve.
plt.plot(fpr, fpr, linestyle = '--', color = 'k')
# We plot a seconary diagonal line, with dashed line style and black color.
plt.xlabel('False positive rate')
plt.ylabel('True positive rate')
plt.title('ROC curve')

In [ ]:
AUROC = roc_auc_score(df_actual_predicted_probs['loan_data_targets_test'], df_actual_predicted_probs['y_hat_test_proba'])
AUROC

## Gini and Kolmogorov-Smirnov

In [ ]:
df_actual_predicted_probs = df_actual_predicted_probs.sort_values('y_hat_test_proba')

In [ ]:
df_actual_predicted_probs.head()

In [ ]:
df_actual_predicted_probs.tail()

In [ ]:
df_actual_predicted_probs = df_actual_predicted_probs.reset_index()

In [ ]:
df_actual_predicted_probs.head()

In [ ]:
df_actual_predicted_probs['Cumulative N Population'] = df_actual_predicted_probs.index + 1
df_actual_predicted_probs['Cumulative N Good'] = df_actual_predicted_probs['loan_data_targets_test'].cumsum()
df_actual_predicted_probs['Cumulative N Bad'] = df_actual_predicted_probs['Cumulative N Population'] - df_actual_predicted_probs['loan_data_targets_test'].cumsum()

In [ ]:
df_actual_predicted_probs.head()

In [ ]:
df_actual_predicted_probs['Cumulative Perc Population'] = df_actual_predicted_probs['Cumulative N Population'] / (df_actual_predicted_probs.shape[0])
df_actual_predicted_probs['Cumulative Perc Good'] = df_actual_predicted_probs['Cumulative N Good'] / df_actual_predicted_probs['loan_data_targets_test'].sum()
df_actual_predicted_probs['Cumulative Perc Bad'] = df_actual_predicted_probs['Cumulative N Bad'] / (df_actual_predicted_probs.shape[0] - df_actual_predicted_probs['loan_data_targets_test'].sum())

In [ ]:
df_actual_predicted_probs.head()

In [ ]:
df_actual_predicted_probs.tail()

In [ ]:
# Plot Gini
plt.plot(df_actual_predicted_probs['Cumulative Perc Population'], df_actual_predicted_probs['Cumulative Perc Bad'])
plt.plot(df_actual_predicted_probs['Cumulative Perc Population'], df_actual_predicted_probs['Cumulative Perc Population'], linestyle = '--', color = 'k')
plt.xlabel('Cumulative % Population')
plt.ylabel('Cumulative % Bad')
plt.title('Gini')

In [ ]:
Gini = AUROC * 2 - 1
Gini

In [ ]:
# Plot KS
plt.plot(df_actual_predicted_probs['y_hat_test_proba'], df_actual_predicted_probs['Cumulative Perc Bad'], color = 'r')
plt.plot(df_actual_predicted_probs['y_hat_test_proba'], df_actual_predicted_probs['Cumulative Perc Good'], color = 'b')
plt.xlabel('Estimated Probability for being Good')
plt.ylabel('Cumulative %')
plt.title('Kolmogorov-Smirnov')

## Applying PD Model
### Calculating the PD of Individual Accounts

In [ ]:
pd.options.display.max_columns = None

In [ ]:
inputs_test_with_ref_cat.head()

In [ ]:
summary_table

Let us look at the first observation from our inputs_test_with_ref_cat table. <br>

**Note that our INTERCEPT = -0.463330**
Note: We will neglect the categories which are used as reference categories. 

The borrower has: 
<li>Grade:C  |  coef_ = 0.547289, 
<li>home_ownership: MORTGAGE  | coef_ = 0.104737, 
<li>Verification Stat:Verified  |  This category is used as a REFERENCE CAT, so the value won't change, i.e. = 0
<li>purpose:major_purch__car__home_impr	  |  coef_ = 0.242061
<li>mths_since_issue_d:40-41   | coef_ = 0.843239
<li>int_rate:12.025-15.74	 |  coef_ = 0.362496
<li>mths_since_earliest_cr_line:165-247   |  Coef_ = 0.057056
<li>inq_last_6mths:0	|   coef_ = 0.265943
<li>annual_inc:60K-70K	 |   Coef_ = 0.121089
<li>dti:7.7-10.5	|   coef_ = 0.134914
<li>mths_since_last_delinq:Missing	|   coef_ = 0.076772
<li>mths_since_last_record:Missing	  |   coef_ = 0.135092


Therefore if we add all these numbers, including the intercept, we get a sum of 2.427358

We will use this number to calculate the Propability of Not Default. 

We know that, ln((1-PD)/PD) = 2.427358
=>  (1-PD)/PD = e(2.427358)
=>  (1-PD)/PD = 12.21653
=>  (1-PD) = 12.21653/(12.21653 + 1)
=>  (1-PD) = 0.924337

We have our answer. Let's check!


In [ ]:
y_hat_test_proba[0]

**Fantastic!**

Our answer matches. However, banks usually create scorecards to study the borrower's profile. 

Scorecards help technical and non-technical people, understand this data and make their decision.<br>
Scorecard have the advantage of being standardized, no matter what model we may use. 

Scorecards look the same way!

Therefore, **we will now create a scorecard for our PD model.**

## Creating a SCORECARD!

In [ ]:
df_ref_categories = pd.DataFrame(ref_categories, columns = ['Feature name'])
df_ref_categories['Coefficients'] = 0
df_ref_categories['p_values'] = np.nan
df_ref_categories

In [ ]:
df_scorecard = pd.concat([summary_table, df_ref_categories])
df_scorecard = df_scorecard.reset_index()
df_scorecard

In [ ]:
df_scorecard['Original feature name'] = df_scorecard['Feature name'].str.split(':').str[0]
df_scorecard

In [ ]:
min_score = 300
max_score = 850

In [ ]:
df_scorecard.groupby('Original feature name')['Coefficients'].min()

In [ ]:
min_sum_coef = df_scorecard.groupby('Original feature name')['Coefficients'].min().sum()
min_sum_coef

In [ ]:
df_scorecard.groupby('Original feature name')['Coefficients'].max()

In [ ]:
max_sum_coef = df_scorecard.groupby('Original feature name')['Coefficients'].max().sum()
max_sum_coef

In [ ]:
#NOw, based on the max and min scores coef, we will create a new column named Score - Calculation
"""
To do so,we multiply the value of the 'Coefficients' column by the ration of the differences between
maximum score and minimum score and maximum sum of coefficients and minimum sum of cefficients.
"""
df_scorecard['Score - Calculation'] = df_scorecard['Coefficients'] * (max_score - min_score) / (max_sum_coef - min_sum_coef)
df_scorecard

In [9]:
df_scorecard['Score - Calculation'][0] = ((df_scorecard['Coefficients'][0] - min_sum_coef) / (max_sum_coef - min_sum_coef)) * (max_score - min_score) + min_score
df_scorecard

NameError: name 'df_scorecard' is not defined

In [ ]:
df_scorecard['Score - Final'] = df_scorecard['Score - Calculation'].round()
df_scorecard

In [ ]:
min_sum_score_prel = df_scorecard.groupby('Original feature name')['Score - Final'].min().sum()
min_sum_score_prel

In [ ]:
max_sum_score_prel = df_scorecard.groupby('Original feature name')['Score - Final'].max().sum()
max_sum_score_prel

## Calculating Credit Score

In [ ]:
inputs_test_with_ref_cat.head()

In [ ]:
df_scorecard

In [ ]:
inputs_test_with_ref_cat_w_intercept = inputs_test_with_ref_cat

In [ ]:
inputs_test_with_ref_cat_w_intercept.insert(0, 'Intercept', 1)

In [ ]:
inputs_test_with_ref_cat_w_intercept.head()

In [ ]:
# We keep only the columns with column names,
# which are exactly equal to the row values of the 'Feature name' column from the 'df_scorecard' dataframe
inputs_test_with_ref_cat_w_intercept = inputs_test_with_ref_cat_w_intercept[df_scorecard['Feature name'].values]

In [ ]:
inputs_test_with_ref_cat_w_intercept.head()

In [ ]:
scorecard_scores = df_scorecard['Score - Final']

In [ ]:
inputs_test_with_ref_cat_w_intercept.shape

In [ ]:
scorecard_scores.shape

In [ ]:
scorecard_scores = scorecard_scores.values.reshape(102, 1)

In [ ]:
#We will perform dot multiplication of both the dfs
y_scores = inputs_test_with_ref_cat_w_intercept.dot(scorecard_scores)

In [ ]:
y_scores.head()

## From Credit Score to PD

In [ ]:
sum_coef_from_score = ((y_scores - min_score) / (max_score - min_score)) * (max_sum_coef - min_sum_coef) + min_sum_coef

In [ ]:
y_hat_proba_from_score = np.exp(sum_coef_from_score) / (np.exp(sum_coef_from_score) + 1)
y_hat_proba_from_score.head()

In [ ]:
y_hat_test_proba[0:5]

In [ ]:
df_actual_predicted_probs['y_hat_test_proba'].head()

## Setting Cut-offs

Now, we need to set a cut-off. Any borrower who comes under this cut off, is good.
Any borrower above this cut-off is bad.
This impacts the quality of loans.

In [ ]:
tr = 0.9
df_actual_predicted_probs['y_hat_test'] = np.where(df_actual_predicted_probs['y_hat_test_proba'] > tr, 1, 0)

In [ ]:
pd.crosstab(df_actual_predicted_probs['loan_data_targets_test'], df_actual_predicted_probs['y_hat_test'], rownames = ['Actual'], colnames = ['Predicted'])

In [ ]:
pd.crosstab(df_actual_predicted_probs['loan_data_targets_test'], df_actual_predicted_probs['y_hat_test'], rownames = ['Actual'], colnames = ['Predicted']) / df_actual_predicted_probs.shape[0]

In [ ]:
(pd.crosstab(df_actual_predicted_probs['loan_data_targets_test'], df_actual_predicted_probs['y_hat_test'], rownames = ['Actual'], colnames = ['Predicted']) / df_actual_predicted_probs.shape[0]).iloc[0, 0] + (pd.crosstab(df_actual_predicted_probs['loan_data_targets_test'], df_actual_predicted_probs['y_hat_test'], rownames = ['Actual'], colnames = ['Predicted']) / df_actual_predicted_probs.shape[0]).iloc[1, 1]

In [ ]:
from sklearn.metrics import roc_curve, roc_auc_score

In [ ]:
roc_curve(df_actual_predicted_probs['loan_data_targets_test'], df_actual_predicted_probs['y_hat_test_proba'])

In [ ]:
fpr, tpr, thresholds = roc_curve(df_actual_predicted_probs['loan_data_targets_test'], df_actual_predicted_probs['y_hat_test_proba'])

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

In [ ]:
plt.plot(fpr, tpr)
plt.plot(fpr, fpr, linestyle = '--', color = 'k')
plt.xlabel('False positive rate')
plt.ylabel('True positive rate')
plt.title('ROC curve')

In [ ]:
thresholds

In [ ]:
thresholds.shape

In [ ]:
df_cutoffs = pd.concat([pd.DataFrame(thresholds), pd.DataFrame(fpr), pd.DataFrame(tpr)], axis = 1)

In [ ]:
df_cutoffs.columns = ['thresholds', 'fpr', 'tpr']

In [ ]:
df_cutoffs.head()

In [ ]:
df_cutoffs['thresholds'][0] = 1 - 1 / np.power(10, 16)

In [ ]:
df_cutoffs['Score'] = ((np.log(df_cutoffs['thresholds'] / (1 - df_cutoffs['thresholds'])) - min_sum_coef) * ((max_score - min_score) / (max_sum_coef - min_sum_coef)) + min_score).round()

In [ ]:
df_cutoffs.head()

In [ ]:
df_cutoffs['Score'][0] = max_score

In [ ]:
df_cutoffs.head()

In [ ]:
df_cutoffs.tail()

In [ ]:
def n_approved(p):
    return np.where(df_actual_predicted_probs['y_hat_test_proba'] >= p, 1, 0).sum()

In [ ]:
# Assuming that all credit applications above a given probability of being 'good' will be approved,
# when we apply the 'n_approved' function to a threshold, it will return the number of approved applications.
# Thus, here we calculate the number of approved appliations for al thresholds.
# Then, we calculate the number of rejected applications for each threshold.
# It is the difference between the total number of applications and the approved applications for that threshold.

df_cutoffs['N Approved'] = df_cutoffs['thresholds'].apply(n_approved)
df_cutoffs['N Rejected'] = df_actual_predicted_probs['y_hat_test_proba'].shape[0] - df_cutoffs['N Approved']
df_cutoffs['Approval Rate'] = df_cutoffs['N Approved'] / df_actual_predicted_probs['y_hat_test_proba'].shape[0]
df_cutoffs['Rejection Rate'] = 1 - df_cutoffs['Approval Rate']

In [ ]:
df_cutoffs.head()

In [ ]:
df_cutoffs.iloc[5000: 6200, ]

In [ ]:
inputs_train_with_ref_cat.to_csv('inputs_train_with_ref_cat.csv')

In [ ]:
df_scorecard.to_csv('df_scorecard.csv')

In [195]:
# Saving the model
import pickle
pickle.dump(reg2, open('pd_model.sav', 'wb'))